muffnn/fm/fm_classifier.py

```diff
@@ -0,0 +1,437 @@
+        Solver to use. If a string is passed, then the corresponding solver
+        from `scipy.optimize.minimize` is used.
+    solver_kwargs : dict, optional
+        Additional keyword arguments to pass to `solver` upon construction.
+        See the TensorFlow documentation for possible options. Typically,
+        one would want to set the `learning_rate`.
+
+    Attributes
+    ----------
+    n_dims_ : int
+        Number of input dimensions.
+    classes_ : array
+        Classes from the data.
+    n_classes_ : int
+        Number of classes.
+    is_sparse_ : bool
+        Whether a model taking sparse input was fit.
+    """
+    def __init__(self, rank=8, batch_size=64, n_epochs=5,
+                 random_state=None, lambda_v=0.0,
+                 lambda_beta=0.0, solver=tf.train.AdadeltaOptimizer,
+                 init_scale=0.1, solver_kwargs=None):
+        self.rank = rank
+        self.batch_size = batch_size
+        self.n_epochs = n_epochs
+        self.random_state = random_state
+        self.lambda_v = lambda_v
+        self.lambda_beta = lambda_beta
+        self.solver = solver
+        self.init_scale = init_scale
+        self.solver_kwargs = solver_kwargs
+
+    def _set_up_graph(self):
+        """Initialize TF objects (needed before fitting or restoring)."""
+
+        # Input values.
+        if self.is_sparse_:
+            self._x_inds = tf.placeholder(tf.int64, [None, 2], "x_inds")
+            self._x_vals = tf.placeholder(tf.float32, [None], "x_vals")
+            self._x_shape = tf.placeholder(tf.int64, [2], "x_shape")
+            self._x = tf.sparse_reorder(
+                tf.SparseTensor(self._x_inds, self._x_vals, self._x_shape))
+            x2 = tf.sparse_reorder(
+                tf.SparseTensor(self._x_inds,
+                                self._x_vals * self._x_vals,
+                                self._x_shape))
+            matmul = tf.sparse_tensor_dense_matmul
+        else:
+            self._x = tf.placeholder(tf.float32, [None, self.n_dims_], "x")
+            x2 = self._x * self._x
+            matmul = tf.matmul
+
+        if self._output_size == 1:
+            self._y = tf.placeholder(tf.float32, [None], "y")
+        else:
+            self._y = tf.placeholder(tf.int32, [None], "y")
+
+        with tf.variable_scope("fm"):
+            self._v = tf.get_variable(
+                "v", [self.rank, self.n_dims_, self._output_size])
+            self._beta = tf.get_variable(
+                "beta", [self.n_dims_, self._output_size])
+            self._beta0 = tf.get_variable("beta0", [self._output_size])
+
+        vx = tf.stack([matmul(self._x, self._v[i, :, :])
+                       for i in range(self.rank)], axis=1)
+        v2 = self._v * self._v
+        v2x2 = tf.stack([matmul(x2, v2[i, :, :])
+                         for i in range(self.rank)], axis=1)
+        int_term = 0.5 * tf.reduce_sum(tf.square(vx) - v2x2, axis=1)
```

Comment:
The `axis=1` is a little confusing since its in the middle of this 3-dim array. Maybe change it to `axis=-1` instead?